ws/api/transport.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-zoo/bone"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/ws"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	protocol = "ws"
+)
+
+var (
+	errUnauthorizedAccess = errors.New("missing or invalid credentials provided")
+	errMalformedSubtopic  = errors.New("malformed subtopic")
+)
+
+var (
+	upgrader = websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+	auth              mainflux.ThingsServiceClient
+	logger            log.Logger
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+)
+
+type subscription struct {
```

Comment:
You have `subscription`, `ws.Client` and `ws.Channel`. Try to unite them to a single `Client`.